AX: live regions should respect alternative text (alt, aria-label, aria-labelledby)

hoffmanjoshua opened the pull request:
#### 2b05daadd0527630a62e38726f33ebfd6663c8d6
<pre>
AX: live regions should respect alternative text (alt, aria-label, aria-labelledby)
<a href="https://bugs.webkit.org/show_bug.cgi?id=303969">https://bugs.webkit.org/show_bug.cgi?id=303969</a>
<a href="https://rdar.apple.com/74236057">rdar://74236057</a>

Reviewed by Tyler Wilcock.

This patch allows for aria-live announcements to start respecting accessible text.

Prior to this patch, we only respected image alt text via the text iterator. But, by switching
to `textUnderElement`, and updating shouldIncludeInSnapshot and textForObject, we can properly
handle alternative text.

The shouldIncludeInSnapshot change is necessary, since alternative text should prevent
textual children from being included in an object&apos;s text.

There are two important changes to `textForObject`:
(1) Most text is computed using textUnderElement rather than using text marker ranges. This
allows us to use the alternative text handling already built into this method (with one
caveat as described in #2).
(2) `textUnderElement` only returns alt. text for descendants of its calling node. This means,
if the object we want text for has an aria-label, that won&apos;t get considered. This is why
we return the description (which accounts for alt text, aria-label, etc.) before using
textUnderElement.

`TextUnderElementMode` has two new options specifically used by live regions:
- includeListMarkers: will return list marker text for list marker objects.
- descendIntoContainers: allows textUnderElement to scoop up text for table, tree, and list
descendants.

Lastly, I added a new parameter, `prependNewline`, to appendNameToStringBuilder. This allows
textUnderElement to insert newlines when appropriate (for example, after list items). To
preserve existing behavior, if we are already inserting a space before an object, we will
not also insert a newline. In the future, we should make `textUnderElement` respect the text
emission behavior of objects, instead of inserting spaces.

Tests: accessibility/mac/live-regions/live-region-aria-label.html
       accessibility/mac/live-regions/live-region-img-alt.html

* LayoutTests/accessibility/mac/live-regions/live-region-aria-label-expected.txt: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-aria-label.html: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-img-alt-expected.txt: Copied from LayoutTests/accessibility/mac/live-regions/live-region-removals-expected.txt.
* LayoutTests/accessibility/mac/live-regions/live-region-img-alt.html: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-removals-expected.txt:
* LayoutTests/accessibility/mac/live-regions/live-region-with-atomic-expected.txt:
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXLiveRegionManager.cpp:
(WebCore::AXLiveRegionManager::buildLiveRegionSnapshot const):
(WebCore::AXLiveRegionManager::shouldIncludeInSnapshot const):
(WebCore::AXLiveRegionManager::textForObject const):
(WebCore::AXLiveRegionManager::computeAnnouncement const):
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::shouldUseAccessibilityObjectInnerText):
(WebCore::appendNameToStringBuilder):
(WebCore::shouldPrependNewline):
(WebCore::AccessibilityNodeObject::textUnderElement const):
(WebCore::accessibleNameForNode):
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::textUnderElement const):

Canonical link: <a href="https://commits.webkit.org/304318@main">https://commits.webkit.org/304318@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fc87ba66d72c8b879a8e9b846da908dfa505f695

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/135194 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7586 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46491 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142700 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86965 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/137063 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/8223 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7435 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103311 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/70514 "Passed tests") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138140 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5864 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/121201 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84169 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5654 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3256 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/3290 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/114858 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/39376 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/145397 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/7269 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39944 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/111690 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7313 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6088 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112054 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28443 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5496 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117495 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/61222 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7323 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35629 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7079 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70874 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7300 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7180 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

minorninth commented on Source/WebCore/accessibility/AXLiveRegionManager.cpp:
```diff
@@ -239,7 +245,19 @@ bool AXLiveRegionManager::shouldIncludeInSnapshot(AccessibilityObject& object) c
+        if (String ariaAlternative = nodeObject->ariaAccessibilityDescription(); ariaAlternative.length())
+            return ariaAlternative;
+
+        if (String altText = nodeObject->altTextFromAttributeOrStyle(); altText.length())
```
I do worry that if we had to check both aria alternatives and alt, we might be missing something. Do we need to worry about the title attribute?

Just wondering if there should already be a helper function somewhere that returns ariaAccessibilityDescription, altTextFromAttributeOrStyle, or anything else similar. Basically we want the accessible name calculation, but not descending into other elements.

hoffmanjoshua replied:
Looking again, it might make sense for us to use the `description()` value here, which also accounts for title?

hoffmanjoshua replied:
Updated the pr to start using `description` here and in `shouldIncludeInSnapshot `

minorninth reviewed (commented):
lgtm

twilco commented on Source/WebCore/accessibility/AXLiveRegionManager.cpp:
```diff
@@ -333,10 +345,12 @@ AttributedString AXLiveRegionManager::computeAnnouncement(const LiveRegionSnapsh
             languageAttribute.set(accessibilityLanguageAttributeKey, AttributedString::AttributeValue { object.language });
-            attributes.append({ { startLocation, object.text.length() }, WTFMove(languageAttribute) });
+            // The - / + 1 allows us to set the language of the space character seemlessly with the text around it.
+            attributes.append({ { needsSpace ? startLocation - 1 : startLocation, needsSpace ? object.text.length() + 1 : object.text.length() }, WTFMove(languageAttribute) });
```
Is it guaranteed that startLocation is always greater than zero when `needsSpace` is true? We subtract one from it, so we need to protect against accidental wrapping

hoffmanjoshua replied:
Updated this to check that startLocation is > 0 in both cases.

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -3730,14 +3730,16 @@ static bool shouldUseAccessibilityObjectInnerText(AccessibilityObject& object, T
-    if (object.isTree() || object.isCanvas())
-        return false;
+        if (object.isTree() || object.isCanvas())
+            return false;
```
All of these blocks just return false. Is there a benefit to having them separated out into groups?

hoffmanjoshua replied:
Merged into one!

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -3747,13 +3749,19 @@ static bool shouldUseAccessibilityObjectInnerText(AccessibilityObject& object, T
         builder.append(' ');
+    else if (prependNewline) {
+        // FIXME: This is in an else if so we don't break existing behavior.
+        // We should update prependSpace to also be based off of textiterator's behavior, so that textUnderElement is more accurate.
```
We should match the casing of TextIterator

Can this comment say more? What is the existing behavior we're trying to not break? What behavior from TextIterator do we want to eventually mirror?

hoffmanjoshua replied:
Expanded on each comment a bit.

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -4205,7 +4219,7 @@ static String accessibleNameForNode(Node& node, Node* labelledbyNode)
     if (axObject) {
         if (axObject->accessibleNameDerivesFromContent())
-            text = axObject->textUnderElement({ TextUnderElementMode::Children::IncludeNameFromContentsChildren, true, true, false, TrimWhitespace::Yes, labelledbyNode });
+            text = axObject->textUnderElement({ TextUnderElementMode::Children::IncludeNameFromContentsChildren, true, true, false, false, false, TrimWhitespace::Yes, labelledbyNode });
```
I wonder if our new parameters should be enums rather than bools. This callsite highlights the issues:

1. It's not clear what these bools are actually doing (they're just true or false)
2. It would be extremely easy to make a mistake and change the wrong bool (e.g. I accidentally flip the include list markers bool to true instead of the bool before it, because none of these have names like an enum would)

hoffmanjoshua committed:
AX: live regions should respect alternative text (alt, aria-label, aria-labelledby)
https://bugs.webkit.org/show_bug.cgi?id=303969
rdar://74236057

Reviewed by Tyler Wilcock.

This patch allows for aria-live announcements to start respecting accessible text.

Prior to this patch, we only respected image alt text via the text iterator. But, by switching
to `textUnderElement`, and updating shouldIncludeInSnapshot and textForObject, we can properly
handle alternative text.

The shouldIncludeInSnapshot change is necessary, since alternative text should prevent
textual children from being included in an object's text.

There are two important changes to `textForObject`:
(1) Most text is computed using textUnderElement rather than using text marker ranges. This
allows us to use the alternative text handling already built into this method (with one
caveat as described in #2).
(2) `textUnderElement` only returns alt. text for descendants of its calling node. This means,
if the object we want text for has an aria-label, that won't get considered. This is why
we return the description (which accounts for alt text, aria-label, etc.) before using
textUnderElement.

`TextUnderElementMode` has two new options specifically used by live regions:
- includeListMarkers: will return list marker text for list marker objects.
- descendIntoContainers: allows textUnderElement to scoop up text for table, tree, and list
descendants.

Lastly, I added a new parameter, `prependNewline`, to appendNameToStringBuilder. This allows
textUnderElement to insert newlines when appropriate (for example, after list items). To
preserve existing behavior, if we are already inserting a space before an object, we will
not also insert a newline. In the future, we should make `textUnderElement` respect the text
emission behavior of objects, instead of inserting spaces.

Tests: accessibility/mac/live-regions/live-region-aria-label.html
       accessibility/mac/live-regions/live-region-img-alt.html

* LayoutTests/accessibility/mac/live-regions/live-region-aria-label-expected.txt: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-aria-label.html: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-img-alt-expected.txt: Copied from LayoutTests/accessibility/mac/live-regions/live-region-removals-expected.txt.
* LayoutTests/accessibility/mac/live-regions/live-region-img-alt.html: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-removals-expected.txt:
* LayoutTests/accessibility/mac/live-regions/live-region-with-atomic-expected.txt:
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXLiveRegionManager.cpp:
(WebCore::AXLiveRegionManager::buildLiveRegionSnapshot const):
(WebCore::AXLiveRegionManager::shouldIncludeInSnapshot const):
(WebCore::AXLiveRegionManager::textForObject const):
(WebCore::AXLiveRegionManager::computeAnnouncement const):
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator<<):
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::shouldUseAccessibilityObjectInnerText):
(WebCore::appendNameToStringBuilder):
(WebCore::shouldPrependNewline):
(WebCore::AccessibilityNodeObject::textUnderElement const):
(WebCore::accessibleNameForNode):
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::textUnderElement const):

Canonical link: https://commits.webkit.org/304318@main
